internal/core/domain/trade_service.go

```diff
@@ -82,226 +86,187 @@ func (t *Trade) Accept(
 	t.Status = AcceptedStatus
 	t.SwapAccept.ID = swapAcceptID
 	t.SwapAccept.Message = swapAcceptMsg
-	t.Timestamp.Accept = uint64(time.Now().Unix())
+	t.SwapAccept.Timestamp = uint64(time.Now().Unix())
 	t.PsetBase64 = psetBase64
-	txID, err := transactionutil.GetTxIdFromPset(psetBase64)
-	if err != nil {
-		return false, nil
-	}
+	txID, _ := PsetManager.GetTxID(psetBase64)
 	t.TxID = txID
 
 	return true, nil
 }
 
-// CompleteResult is return type of Complete method
+// CompleteResult is return type of Complete method.
 type CompleteResult struct {
 	OK    bool
 	TxHex string
 	TxID  string
 }
 
-// Complete sets the status of the trade to Complete by adding the txID
-// of the tx in the blockchain. The trade must be in Accepted or
-// FailedToComplete status for being completed, otherwise an error is thrown
+// Complete brings a trade from the Accepted to the Completed status by
+// checiking that the given PSET completes the one of the SwapAccept message
+// and by finalizing it and extracting the raw tx in hex format.
 func (t *Trade) Complete(psetBase64 string) (*CompleteResult, error) {
-	if t.IsCompleted() {
+	if t.Status.Code >= Completed {
 		return &CompleteResult{OK: true, TxHex: t.TxHex, TxID: t.TxID}, nil
 	}
 
-	if !t.IsAccepted() {
-		return nil, ErrMustBeAccepted
-	}
-
-	if err := pkgswap.ValidateCompletePset(pkgswap.ValidateCompletePsetOpts{
-		PsetBase64:         psetBase64,
-		InputBlindingKeys:  t.SwapAcceptMessage().GetInputBlindingKey(),
-		OutputBlindingKeys: t.SwapAcceptMessage().GetOutputBlindingKey(),
-		SwapRequest:        t.SwapRequestMessage(),
-	}); err != nil {
-		t.Fail(
-			t.SwapAccept.ID,
-			FailedToCompleteStatus,
-			pkgswap.ErrCodeFailedToComplete,
-			err.Error(),
-		)
-		return &CompleteResult{OK: false}, nil
+	if t.IsExpired() {
+		if !t.Status.Failed {
```

Comment:
why we would prevent to write again Failed?

Reply:
this just make sure that the trade is set to fail status if expired. The check just prevent setting the status to Failed if it already is.